File: pickerkit/src/main/java/com/dewarder/pickerkit/panel/PickerPanelCategoryAdapter.java
package com.dewarder.pickerkit.panel;

import android.support.annotation.IdRes;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.dewarder.pickerkit.R;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class PickerPanelCategoryAdapter extends RecyclerView.Adapter<PickerPanelCategoryViewHolder> {

    private final ArrayList<PickerCategory> mCategories = new ArrayList<>();
    private OnAttachmentPanelCategoryClickListener mOnAttachmentPanelCategoryClickListener;

    @Override
    public PickerPanelCategoryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_attachment_category, parent, false);
        return new PickerPanelCategoryViewHolder(view);
    }

    @Override
    public void onBindViewHolder(PickerPanelCategoryViewHolder holder, int position) {
        PickerCategory item = mCategories.get(position);
        holder.itemView.setId(item.getId());
        holder.setName(item.getName());
        holder.setCircleIcon(item.getIcon());
        holder.setCircleBackground(new CircleDrawable(item.getColor()));
        holder.itemView.setOnClickListener(v -> notifyCategoryClicked(item));
    }

    @Override
    public int getItemCount() {
        return mCategories.size();
    }

    public void setCategories(List<PickerCategory> categories) {
        mCategories.clear();
        mCategories.addAll(categories);
        notifyDataSetChanged();
    }

    public void add(PickerCategory category) {
        mCategories.add(category);
        notifyDataSetChanged();
    }

    public void addAll(Collection<PickerCategory> categories) {
        mCategories.addAll(categories);
        notifyDataSetChanged();
    }

    public void replace(@IdRes int oldCategoryId, PickerCategory newCategory) {
        int index = findCategoryById(oldCategoryId);
        mCategories.set(index, newCategory);
        notifyItemChanged(index);
    }

    public void remove(@IdRes int categoryId) {
        int index = findCategoryById(categoryId);
        mCategories.remove(index);
        notifyItemRemoved(index);
    }

    private int findCategoryById(@IdRes int id) {
        int index = -1;
        for (int i = 0; i < mCategories.size(); i++) {
            if (mCategories.get(i).getId() == id) {
                index = i;
                break;
            }
        }

        if (index == -1) {
            throw new IllegalStateException("Category with id " + id + " isn't found");
        } else {
            return index;
        }
    }

    public void setOnAttachmentPanelCategoryClickListener(OnAttachmentPanelCategoryClickListener listener) {
        mOnAttachmentPanelCategoryClickListener = listener;
    }

    private void notifyCategoryClicked(PickerCategory item) {
        if (mOnAttachmentPanelCategoryClickListener != null) {
            mOnAttachmentPanelCategoryClickListener.onPanelPickerClicked(item.getId());
        }
    }
}

File: pickerkit/src/main/java/com/dewarder/pickerkit/panel/PickerPanelCategoryAdapter.kt
package com.dewarder.pickerkit.panel

import android.support.annotation.IdRes
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.dewarder.pickerkit.R

import java.util.ArrayList

class PickerPanelCategoryAdapter : RecyclerView.Adapter<PickerPanelCategoryViewHolder>() {

    private val mCategories = ArrayList<PickerCategory>()
    private var mOnPickerPanelCategoryClickListener: OnPickerPanelCategoryClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PickerPanelCategoryViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_attachment_category, parent, false)
        return PickerPanelCategoryViewHolder(view)
    }

    override fun onBindViewHolder(holder: PickerPanelCategoryViewHolder, position: Int) {
        val item = mCategories[position]
        holder.itemView.id = item.id
        holder.setName(item.name)
        holder.setCircleIcon(item.icon)
        holder.setCircleBackground(CircleDrawable(item.color))
        holder.itemView.setOnClickListener { v -> notifyCategoryClicked(item) }
    }

    override fun getItemCount(): Int {
        return mCategories.size
    }

    fun setCategories(categories: List<PickerCategory>) {
        mCategories.clear()
        mCategories.addAll(categories)
        notifyDataSetChanged()
    }

    fun add(category: PickerCategory) {
        mCategories.add(category)
        notifyDataSetChanged()
    }

    fun addAll(categories: Collection<PickerCategory>) {
        mCategories.addAll(categories)
        notifyDataSetChanged()
    }

    fun replace(@IdRes oldCategoryId: Int, newCategory: PickerCategory) {
        val index = findCategoryById(oldCategoryId)
        mCategories[index] = newCategory
        notifyItemChanged(index)
    }

    fun remove(@IdRes categoryId: Int) {
        val index = findCategoryById(categoryId)
        mCategories.removeAt(index)
        notifyItemRemoved(index)
    }

    private fun findCategoryById(@IdRes id: Int): Int {
        var index = -1
        for (i in mCategories.indices) {
            if (mCategories[i].id == id) {
                index = i
                break
            }
        }

        return if (index == -1) {
            throw IllegalStateException("Category with id $id isn't found")
        } else {
            index
        }
    }

    fun setOnAttachmentPanelCategoryClickListener(listener: OnPickerPanelCategoryClickListener) {
        mOnPickerPanelCategoryClickListener = listener
    }

    private fun notifyCategoryClicked(item: PickerCategory) {
        if (mOnPickerPanelCategoryClickListener != null) {
            mOnPickerPanelCategoryClickListener!!.onPickerCategoryClicked(item.id)
        }
    }
}
